### ISSUE TITLE
Time taken for Controlled ODEs varies a lot with Initial State

### ISSUE THREAD

**Issue by AnikenC** [patrick-kidger/diffrax#263]
Hello! Diffrax has been an incredible package for DEs, and I'm very impressed with all the tools available, however I've been getting a weird issue when applying to my project. I'm noticing that based on the initial states of a Coupled ODE system, the time taken for the solver can change by a factor of 25 times, and I'm unable to bring this down any further. I've been trying to solve a system of ODEs (Coherent State Langevin Equations of a Resonator and Transmon Qubit from QM), and below is an example of the code.

The system is two Complex-valued Coupled ODEs which are directly controlled by two input drives. This simulation is the base for training a RL agent, and so part of the initial state, the two drives (controls), and two frequency values for the drive are all variable parameters. For each run, I need to simulate when the transmon is in its ground state (initial state is 0) and excited state (initial state is 1) from which info is sent back to the RL model. Since I would have 100 runs in a batch on the same policy, I've tried using vmap but the time taken is still between 7.8 to 8.5ms.

The primary issue I'm noticing is that when I only simulate in the ground state, the simulation takes on the order of 200 to 300us which is perfect! But when I simulate in the excited state, the simulation takes about 7500 to 7800us, which added together gives the overall time I measure. I'm not too sure why this is occurring, as when I plot the results in the two cases (with the second code section), it doesn't appear that the solution is very stiff. Even so, I've played around with using Explicit (Tsit5) and Implicit (Kvaerno5) solvers, but Tsit5 remained the fastest. I've also checked the steps taken for each case, and it does seem that the steps taken correlates with the time taken (excited case takes 5000 steps, ground takes 190 steps). I've also played around with the initial dt0 and the PIDController params, however after tuning this is the best performance I get. Is there any better way to solve the excited state case and reduce steps taken with Diffrax? While 8ms is still decently fast, the extra order of magnitude in time taken is a big pain when I have to repeat the simulation for 200k to 300k steps for the RL training.

```python
import time

import jax.numpy as jnp
from jax import jit, config, vmap, block_until_ready
from diffrax import (
    diffeqsolve,
    ODETerm,
    Tsit5,
    Kvaerno5,
    LinearInterpolation,
    SaveAt,
    PIDController,
)

from jaxtyping import Array

config.update("jax_enable_x64", True)
out_float_dtype = jnp.float64
out_complex_dtype = jnp.complex128

TS_ACTIONS = jnp.linspace(0.0, 1.0, 51, dtype=out_float_dtype)
TS_SIM = jnp.linspace(0.0, 1.0, 201, dtype=out_float_dtype)


def single_eval(
    res_state: Array,
    res_drive: Array,
    trans_drive: Array,
    res_freq: Array,
    trans_freq: Array,
):
    float_dtype = jnp.float64
    complex_dtype = jnp.complex128
    state_g = jnp.array([*res_state, 0.0, 0.0], dtype=float_dtype)
    state_e = jnp.array([*res_state, 1.0, 0.0], dtype=float_dtype)
    ts_actions = jnp.linspace(0.0, 1.0, 51, dtype=float_dtype)
    ts_sim = jnp.linspace(0.0, 1.0, 201, dtype=float_dtype)

    dt0 = 1e-3
    solver = Tsit5()
    max_steps = int(16**4)
    stepsize_controller = PIDController(
        rtol=1e-4, atol=1e-7, pcoeff=0.3, icoeff=0.4, dcoeff=0.1, jump_ts=ts_actions
    )
    saveat = SaveAt(ts=ts_sim, fn=lambda t, y, args: y[0] + 1j * y[1])

    drive_arr = jnp.vstack((res_drive, trans_drive), dtype=complex_dtype).T
    control = LinearInterpolation(ts=ts_actions, ys=drive_arr)

    args = jnp.array(
        [
            44400.38147835797 - 2.6677954118222127j,
            30070.71498952447 - 0.019959690218603887j,
            -0.013361622621663578,
            -9.792069634763914,
            -1795.9757810978654,
            0.0013124927820996545,
            0.48105931278166897,
            58.82111534791434,
            0.05202123087967161,
            res_freq[0],
            trans_freq[0],
        ],
        dtype=complex_dtype,
    )

    def vector_field(t, y, args):
        cr, ci, qr, qi = y
        drive_res, drive_trans = control.evaluate(t)
        (
            wc_eff,
            wq_eff,
            k1,
            k2,
            k3,
            a1,
            a2,
            a3,
            a4,
            res_drive_freq,
            trans_drive_freq,
        ) = args
        c = cr + 1j * ci
        q = qr + 1j * qi
        c_squared = cr**2 + ci**2
        q_squared = qr**2 + qi**2
        c_freq_term = drive_res * jnp.exp(-1j * (res_drive_freq - wc_eff.real) * t)
        q_freq_term = drive_trans * jnp.exp(-1j * (trans_drive_freq - wq_eff.real) * t)
        d_c = -1j * (
            (
                wc_eff.imag * 1j
                + c_squared * (k1 + 2 * a1 * q_squared)
                + q_squared * (k2 + a2 * q_squared)
            )
            * c
            + c_freq_term
            + a4 * q_freq_term
        )
        d_q = -1j * (
            (
                wq_eff.imag * 1j
                + c_squared * (k2 + a1 * c_squared + 2 * a2 * q_squared)
                + q_squared * (a3 * q_squared + k3)
            )
            * q
            + q_freq_term
            - a4 * c_freq_term
        )
        return jnp.array([d_c.real, d_c.imag, d_q.real, d_q.imag], dtype=float_dtype)

    ode_term = ODETerm(vector_field)

    sol_g = diffeqsolve(
        terms=ode_term,
        solver=solver,
        t0=0.0,
        t1=1.0,
        dt0=dt0,
        y0=state_g,
        args=args,
        saveat=saveat,
        stepsize_controller=stepsize_controller,
        max_steps=max_steps,
    )

    sol_e = diffeqsolve(
        terms=ode_term,
        solver=solver,
        t0=0.0,
        t1=1.0,
        dt0=dt0,
        y0=state_e,
        args=args,
        saveat=saveat,
        stepsize_controller=stepsize_controller,
        max_steps=max_steps,
    )
    return sol_g.ys, sol_e.ys, sol_g.stats, sol_e.stats


in_state = jnp.array([0.0, 0.0], dtype=out_float_dtype)
in_drive_res = jnp.zeros_like(TS_ACTIONS, dtype=out_complex_dtype) + 10.0
in_drive_trans = jnp.zeros_like(TS_ACTIONS, dtype=out_complex_dtype) + 0.5
in_res_freq = jnp.array([44400.38147835797], dtype=out_float_dtype)
in_trans_freq = jnp.array([30070.71498952447], dtype=out_complex_dtype)

jitted_single = jit(single_eval)

n_trials = 100
b_state = jnp.zeros((2, n_trials), dtype=out_float_dtype)
b_drive_res = jnp.zeros((51, n_trials), dtype=out_complex_dtype) + 10.0
b_drive_trans = jnp.zeros((51, n_trials), dtype=out_complex_dtype) + 0.5
b_res_freq = jnp.zeros((1, n_trials), dtype=out_float_dtype) + 44400.38147835797
b_trans_freq = jnp.zeros((1, n_trials), dtype=out_complex_dtype) + 30070.71498952447

single_g, single_e, g_stats, e_stats = jitted_single(
    in_state, in_drive_res, in_drive_trans, in_res_freq, in_trans_freq
)

start = time.time()
second_g, second_e, g_stats, e_stats = block_until_ready(
    jitted_single(in_state, in_drive_res, in_drive_trans, in_res_freq, in_trans_freq)
)
now = time.time()

batched_eval = jit(vmap(single_eval, in_axes=1))
batched_, batched__, a_, b_ = batched_eval(
    b_state, b_drive_res, b_drive_trans, b_res_freq, b_trans_freq
)
print(f"time taken for single g + e: {(now - start)*1e3}ms")
print(f"for g")
print(f"max_steps: {g_stats['max_steps']}")
print(f"num_accepted_steps: {g_stats['num_accepted_steps']}")
print(f"num_rejected_steps: {g_stats['num_rejected_steps']}")
print(f"num_steps: {g_stats['num_steps']}")
print(f"for e")
print(f"max_steps: {e_stats['max_steps']}")
print(f"num_accepted_steps: {e_stats['num_accepted_steps']}")
print(f"num_rejected_steps: {e_stats['num_rejected_steps']}")
print(f"num_steps: {e_stats['num_steps']}")

start = time.time()
batched_g, batched_e, batched_g_stats, batched_e_stats = block_until_ready(
    batched_eval(b_state, b_drive_res, b_drive_trans, b_res_freq, b_trans_freq)
)

now = time.time()
print(f"time taken for single g + e after batching: {(now - start)/n_trials*1e3}ms")
print(f"for g")
print(f"max_steps: {batched_g_stats['max_steps'][0]}")
print(f"num_accepted_steps: {batched_g_stats['num_accepted_steps'][0]}")
print(f"num_rejected_steps: {batched_g_stats['num_rejected_steps'][0]}")
print(f"num_steps: {batched_g_stats['num_steps'][0]}")
print(f"for e")
print(f"max_steps: {batched_e_stats['max_steps'][0]}")
print(f"num_accepted_steps: {batched_e_stats['num_accepted_steps'][0]}")
print(f"num_rejected_steps: {batched_e_stats['num_rejected_steps'][0]}")
print(f"num_steps: {batched_e_stats['num_steps'][0]}")
 
```

```
def simple_plotter_debug(ts, complex_1, complex_2, name_1, name_2, fig_name):
    fig, ax = plt.subplots(4, 2, num=fig_name)
    ax[0, 0].plot(ts, jnp.absolute(complex_1), label=f"{name_1} photons", color="red")
    ax[0, 0].legend()

    ax[1, 0].plot(
        complex_1.real, complex_1.imag, label=f"{name_1} phase", color="orange"
    )
    ax[1, 0].legend()

    ax[2, 0].plot(ts, complex_1.real, label=f"{name_1} real", color="blue")
    ax[2, 0].legend()

    ax[3, 0].plot(ts, complex_1.imag, label=f"{name_1} imag", color="green")
    ax[3, 0].legend()

    ax[0, 1].plot(ts, jnp.absolute(complex_2), label=f"{name_2} photons", color="red")
    ax[0, 1].legend()

    ax[1, 1].plot(
        complex_2.real, complex_2.imag, label=f"{name_2} phase", color="orange"
    )
    ax[1, 1].legend()

    ax[2, 1].plot(ts, complex_2.real, label=f"{name_2} real", color="blue")
    ax[2, 1].legend()

    ax[3, 1].plot(ts, complex_2.imag, label=f"{name_2} imag", color="green")
    ax[3, 1].legend()

    return fig, ax

import matplotlib.pyplot as plt

batch_g_element = batched_g[0]
batch_e_element = batched_e[0]

res_g = batch_g_element
res_e = batch_e_element
fig1, ax1 = simple_plotter_debug(
    ts=TS_SIM, complex_1=res_g, complex_2=res_e, name_1="g", name_2="e", fig_name="Res"
)
plt.show()
```

**Comment by AnikenC** [comment]
Apologies, I had been working with diffrax with a few modifications to handle complex dtypes as per Issue #197. Below is a working version for the current diffrax version with no modifications; I still get the same code time issues in it where the initial excited state takes far larger than steps than expected.

``` python
import time
import matplotlib.pyplot as plt

import jax.numpy as jnp
from jax import jit, config, vmap, block_until_ready
from diffrax import (
    diffeqsolve,
    ODETerm,
    Tsit5,
    Kvaerno5,
    LinearInterpolation,
    SaveAt,
    PIDController,
)

from jaxtyping import Array

config.update("jax_enable_x64", True)
out_float_dtype = jnp.float64
out_complex_dtype = jnp.complex128

TS_ACTIONS = jnp.linspace(0.0, 1.0, 51, dtype=out_float_dtype)
TS_SIM = jnp.linspace(0.0, 1.0, 201, dtype=out_float_dtype)


def single_eval(
    res_state: Array,
    res_drive: Array,
    trans_drive: Array,
    res_freq: Array,
    trans_freq: Array,
):
    float_dtype = jnp.float64
    complex_dtype = jnp.complex128
    state_g = jnp.array([*res_state, 0.0, 0.0], dtype=float_dtype)
    state_e = jnp.array([*res_state, 1.0, 0.0], dtype=float_dtype)
    ts_actions = jnp.linspace(0.0, 1.0, 51, dtype=float_dtype)
    ts_sim = jnp.linspace(0.0, 1.0, 201, dtype=float_dtype)

    dt0 = 1e-3
    solver = Tsit5()
    max_steps = int(16**4)
    stepsize_controller = PIDController(
        rtol=1e-4, atol=1e-7, pcoeff=0.3, icoeff=0.4, dcoeff=0.1, jump_ts=ts_actions
    )
    saveat = SaveAt(
        ts=ts_sim, fn=lambda t, y, args: jnp.array([y[0], y[1]], dtype=float_dtype)
    )

    drive_arr = jnp.vstack((res_drive, trans_drive), dtype=complex_dtype).T
    control = LinearInterpolation(ts=ts_actions, ys=drive_arr)

    args = jnp.array(
        [
            44400.38147835797 - 2.6677954118222127j,
            30070.71498952447 - 0.019959690218603887j,
            -0.013361622621663578,
            -9.792069634763914,
            -1795.9757810978654,
            0.0013124927820996545,
            0.48105931278166897,
            58.82111534791434,
            0.05202123087967161,
            res_freq[0],
            trans_freq[0],
        ],
        dtype=complex_dtype,
    )

    def vector_field(t, y, args):
        cr, ci, qr, qi = y
        drive_res, drive_trans = control.evaluate(t)
        (
            wc_eff,
            wq_eff,
            k1,
            k2,
            k3,
            a1,
            a2,
            a3,
            a4,
            res_drive_freq,
            trans_drive_freq,
        ) = args
        c = cr + 1j * ci
        q = qr + 1j * qi
        c_squared = cr**2 + ci**2
        q_squared = qr**2 + qi**2
        c_freq_term = drive_res * jnp.exp(-1j * (res_drive_freq - wc_eff.real) * t)
        q_freq_term = drive_trans * jnp.exp(-1j * (trans_drive_freq - wq_eff.real) * t)
        d_c = -1j * (
            (
                wc_eff.imag * 1j
                + c_squared * (k1 + 2 * a1 * q_squared)
                + q_squared * (k2 + a2 * q_squared)
            )
            * c
            + c_freq_term
            + a4 * q_freq_term
        )
        d_q = -1j * (
            (
                wq_eff.imag * 1j
                + c_squared * (k2 + a1 * c_squared + 2 * a2 * q_squared)
                + q_squared * (a3 * q_squared + k3)
            )
            * q
            + q_freq_term
            - a4 * c_freq_term
        )
        return jnp.array([d_c.real, d_c.imag, d_q.real, d_q.imag], dtype=float_dtype)

    ode_term = ODETerm(vector_field)

    sol_g = diffeqsolve(
        terms=ode_term,
        solver=solver,
        t0=0.0,
        t1=1.0,
        dt0=dt0,
        y0=state_g,
        args=args,
        saveat=saveat,
        stepsize_controller=stepsize_controller,
        max_steps=max_steps,
    )

    sol_e = diffeqsolve(
        terms=ode_term,
        solver=solver,
        t0=0.0,
        t1=1.0,
        dt0=dt0,
        y0=state_e,
        args=args,
        saveat=saveat,
        stepsize_controller=stepsize_controller,
        max_steps=max_steps,
    )
    return sol_g.ys, sol_e.ys, sol_g.stats, sol_e.stats


in_state = jnp.array([0.0, 0.0], dtype=out_float_dtype)
in_drive_res = jnp.zeros_like(TS_ACTIONS, dtype=out_complex_dtype) + 10.0
in_drive_trans = jnp.zeros_like(TS_ACTIONS, dtype=out_complex_dtype) + 0.5
in_res_freq = jnp.array([44400.38147835797], dtype=out_float_dtype)
in_trans_freq = jnp.array([30070.71498952447], dtype=out_float_dtype)

jitted_single = jit(single_eval)

n_trials = 100
b_state = jnp.zeros((2, n_trials), dtype=out_float_dtype)
b_drive_res = jnp.zeros((51, n_trials), dtype=out_complex_dtype) + 10.0
b_drive_trans = jnp.zeros((51, n_trials), dtype=out_complex_dtype) + 0.5
b_res_freq = jnp.zeros((1, n_trials), dtype=out_float_dtype) + 44400.38147835797
b_trans_freq = jnp.zeros((1, n_trials), dtype=out_float_dtype) + 30070.71498952447

single_g, single_e, g_stats, e_stats = jitted_single(
    in_state, in_drive_res, in_drive_trans, in_res_freq, in_trans_freq
)

start = time.time()
second_g, second_e, g_stats, e_stats = block_until_ready(
    jitted_single(in_state, in_drive_res, in_drive_trans, in_res_freq, in_trans_freq)
)
now = time.time()

batched_eval = jit(vmap(single_eval, in_axes=1))
batched_, batched__, a_, b_ = batched_eval(
    b_state, b_drive_res, b_drive_trans, b_res_freq, b_trans_freq
)
print(f"time taken for single g + e: {(now - start)*1e3}ms")
print(f"for g")
print(f"max_steps: {g_stats['max_steps']}")
print(f"num_accepted_steps: {g_stats['num_accepted_steps']}")
print(f"num_rejected_steps: {g_stats['num_rejected_steps']}")
print(f"num_steps: {g_stats['num_steps']}")
print(f"for e")
print(f"max_steps: {e_stats['max_steps']}")
print(f"num_accepted_steps: {e_stats['num_accepted_steps']}")
print(f"num_rejected_steps: {e_stats['num_rejected_steps']}")
print(f"num_steps: {e_stats['num_steps']}")

start = time.time()
batched_g, batched_e, batched_g_stats, batched_e_stats = block_until_ready(
    batched_eval(b_state, b_drive_res, b_drive_trans, b_res_freq, b_trans_freq)
)

now = time.time()
print(f"time taken for single g + e after batching: {(now - start)/n_trials*1e3}ms")
print(f"for g")
print(f"max_steps: {batched_g_stats['max_steps'][0]}")
print(f"num_accepted_steps: {batched_g_stats['num_accepted_steps'][0]}")
print(f"num_rejected_steps: {batched_g_stats['num_rejected_steps'][0]}")
print(f"num_steps: {batched_g_stats['num_steps'][0]}")
print(f"for e")
print(f"max_steps: {batched_e_stats['max_steps'][0]}")
print(f"num_accepted_steps: {batched_e_stats['num_accepted_steps'][0]}")
print(f"num_rejected_steps: {batched_e_stats['num_rejected_steps'][0]}")
print(f"num_steps: {batched_e_stats['num_steps'][0]}")


def simple_plotter_debug(ts, complex_1, complex_2, name_1, name_2, fig_name):
    fig, ax = plt.subplots(4, 2, num=fig_name)
    ax[0, 0].plot(ts, jnp.absolute(complex_1), label=f"{name_1} photons", color="red")
    ax[0, 0].legend()

    ax[1, 0].plot(
        complex_1.real, complex_1.imag, label=f"{name_1} phase", color="orange"
    )
    ax[1, 0].legend()

    ax[2, 0].plot(ts, complex_1.real, label=f"{name_1} real", color="blue")
    ax[2, 0].legend()

    ax[3, 0].plot(ts, complex_1.imag, label=f"{name_1} imag", color="green")
    ax[3, 0].legend()

    ax[0, 1].plot(ts, jnp.absolute(complex_2), label=f"{name_2} photons", color="red")
    ax[0, 1].legend()

    ax[1, 1].plot(
        complex_2.real, complex_2.imag, label=f"{name_2} phase", color="orange"
    )
    ax[1, 1].legend()

    ax[2, 1].plot(ts, complex_2.real, label=f"{name_2} real", color="blue")
    ax[2, 1].legend()

    ax[3, 1].plot(ts, complex_2.imag, label=f"{name_2} imag", color="green")
    ax[3, 1].legend()

    return fig, ax


batch_g_element = batched_g[0]
batch_e_element = batched_e[0]

res_g = batch_g_element[:, 0] + 1j * batch_g_element[:, 1]
res_e = batch_e_element[:, 0] + 1j * batch_e_element[:, 1]
fig1, ax1 = simple_plotter_debug(
    ts=TS_SIM, complex_1=res_g, complex_2=res_e, name_1="g", name_2="e", fig_name="Res"
)
plt.show()

```

**Comment by patrick-kidger** [comment]
I'm afraid this code snippet is larger than I can reasonably help you with! But FWIW it looks like you're doing things right -- your jit'ing, using `block_until_ready`, are using `jump_ts` alongside a linear control, etc.

It's not uncommon for the initial conditions to affect things substantially, and that might simply be how things are.

The only real suggestion I've got is to try using `SaveAt(steps=True)`, and checking where all these extra steps are being taken. Is there some unusual pattern to them? What is the value of the vector field at these points -- again, check if there anything unexpected.

**Comment by AnikenC** [comment]
Oh my goodness, thanks for the suggestion! On checking the exact values of the vector field at the time steps taken by the solver, I noticed that I had some very high-frequency components in the equation. Turns out that I had added some extra terms which weren't necessary for the system I'm trying to simulate. Now with the corrections, the simulation takes 300us for simulating both the ground and excited state, and the steps taken are between 50 to 120 for the two states, which is much nicer to know!
Thanks for the advice!
